pymeasure/instruments/yokogawa/aq6370d.py

```diff
@@ -0,0 +1,208 @@
+#
+# This file is part of the PyMeasure package.
+#
+# Copyright (c) 2013-2024 PyMeasure Developers
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+import logging
+
+from pymeasure.instruments import Instrument
+from pymeasure.instruments.validators import strict_discrete_set, strict_range
+
+log = logging.getLogger(__name__)
+log.addHandler(logging.NullHandler())
+
+
+class AQ6370D(Instrument):
```

Comment:
Yes, you are right, it works for the mentioned types of AQ6370 Series (have to check the others like 6370E, 6374, etc)!
I'm however unsure about how I should implement the subclasses then. They all vary with regards to resolution, wavelength range they cover... should I let them inherit it from the class AQ6370 and add them at the bottom (like in commit 091bea8f22d427afde3577bc4fb9f6515bf14a52, however still without the instrument specific ranges, have to look at all the manuals still) or should I create them as separate instruments with their respective files?

Reply:
> They all vary with regards to resolution, wavelength range they cover

Ah, I see.

The different ranges are necessary for the validators, aren't they?
- in this case, you could make the properties _dynamic properties_ and adjust the ranges for every device class (you do not have to add the other devices yourself).
- you could also (I guess there are appropriate commands) read the limits from the device and set the `values` of the properties (which have to be dynamic as well) automatically.
- You could remove the validators altogether and check for errors after setting (`check_set_errors=True`) such that the device itself will complain if the values are wrong. Therefore, you do not have to adjust the code as each device knows its own limits best.
- The most simple course of action right now is to keep the class name as it is right now and add a sentence in the docstring, that the commands work for all devices of that series (maybe mentioning the names), but with different limits.

Depending on your coding abilities, interest, and time, you can choose any of these paths.
Obviously it is better to make it work for more instruments at once, but that can be also a second step.

Reply:
I just added your "quick to add" suggestions like using "A" for trace input, initiating & getting the analysis. Will need to update the test, too.
I'll get back to the other things next week, and thanks for your quick feedback!

Reply:
I just committed it as aq6370series (https://github.com/pymeasure/pymeasure/pull/1059/commits/ad26b1ea3e555da1c12a0a507e45656af2a71a17) and then added the subclasses for the respective instruments (AQ6370C, AQ6370D, AQ6373, AQ6373B, AQ6375 and AQ6375B) and their ranges. But I kept the test for the specific AQ6370D since that's what we have here and with which we generated the test.

Only difference with our given functionalities between the B variant of the instrument is that the B variant can adjust sweep speed, so I have the class aq6373B depended on aq6373, I hope that's fine!

I also didn't add the (discontinued) AQ6370 itself since I'm unsure about whether it would work like this because the remote control manual is specific for the mentioned instruments.

Reply:
> But I kept the test for the specific AQ6370D since that's what we have here and with which we generated the test.

It is ok to keep the test specific, but I'd change the filename to match the filename of the file under test.

> I also didn't add the (discontinued) AQ6370 itself since I'm unsure about whether it would work like this because the remote control manual is specific for the mentioned instruments.

It is good to not add it.

> Only difference with our given functionalities between the B variant of the instrument is that the B variant can adjust sweep speed, so I have the class aq6373B depended on aq6373, I hope that's fine!

Inheriting from other instruments is fine, as we do not need to duplicate code needlessly.